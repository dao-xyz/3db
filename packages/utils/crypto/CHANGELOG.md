# Changelog

## [1.0.4](https://github.com/dao-xyz/peerbit/compare/@peerbit/crypto-v1.0.3...@peerbit/crypto-v1.0.4) (2023-06-15)


### Bug Fixes

* bump dependencies ([8a8fd44](https://github.com/dao-xyz/peerbit/commit/8a8fd440149a966337382db77afe1071141e5c74))


### Dependencies

* The following workspace dependencies were updated
  * dependencies
    * @dao-xyz/uint8arrays bumped from 0.0.4 to 1.0.0

## [1.0.3](https://github.com/dao-xyz/peerbit/compare/@peerbit/crypto-v1.0.0-alpha1...@peerbit/crypto-v1.0.3) (2023-06-14)


### Bug Fixes

* update invalid versions from prerelease release-please ([e2f6411](https://github.com/dao-xyz/peerbit/commit/e2f6411d46edf6d36723ca1ea81d1e55a09d3cd4))

## 1.0.0-alpha1 (2023-06-14)


### âš  BREAKING CHANGES

* only allow Ed25519 PeerIds
* simplified identity handling

### Features

* only allow Ed25519 PeerIds ([532c8b3](https://github.com/dao-xyz/peerbit/commit/532c8b35bc4e85719669db47639ec5ffd11c8eab))
* simplified identity handling ([1ae2416](https://github.com/dao-xyz/peerbit/commit/1ae24168a5c8629b8f9d1c57eceed6abd4a15020))

## @peerbit/crypto [0.1.16](https://github.com/dao-xyz/peerbit/compare/@peerbit/crypto@0.1.15...@peerbit/crypto@0.1.16) (2023-06-07)


### Bug Fixes

* add release cfg ([de76654](https://github.com/dao-xyz/peerbit/commit/de766548f8106804d319e8b51e9607f2a3f60726))





### Dependencies

* **@dao-xyz/uint8arrays:** upgraded to 0.0.4
