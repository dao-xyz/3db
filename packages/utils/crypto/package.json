{
	"name": "@peerbit/crypto",
	"version": "2.1.4",
	"description": "Crypto fn",
	"type": "module",
	"sideEffects": false,
	"module": "lib/esm/index.js",
	"types": "lib/esm/index.d.ts",
	"exports": {
		"import": "./lib/esm/index.js"
	},
	"files": [
		"lib",
		"src",
		"!src/**/__tests__",
		"!lib/**/__tests__",
		"!src/**/__benchmark__",
		"!lib/**/__benchmark__",
		"LICENSE"
	],
	"publishConfig": {
		"access": "public"
	},
	"browser": {
		"./lib/esm/hash.js": "./lib/esm/hash-browser.js",
		"./hash.js": "./lib/esm/hash-browser.js",
		"./lib/esm/random.js": "./lib/esm/random-browser.js",
		"./random.js": "./lib/esm/random-browser.js",
		"./lib/esm/ed25519-sign.js": "./lib/esm/ed25519-sign-browser.js",
		"./ed25519-sign.js": "./lib/esm/ed25519-sign-browser.js"
	},
	"scripts": {
		"clean": "shx rm -rf lib/*",
		"build": "yarn clean && tsc -p tsconfig.json",
		"test": "node ../../../node_modules/.bin/jest test  -c  ../../../jest.config.ts --runInBand --forceExit",
		"test:unit": "node ../../../node_modules/.bin/jest test  -c  ../../../jest.config.unit.ts --runInBand --forceExit",
		"test:integration": "node ../node_modules/.bin/jest test -c  ../../../jest.config.integration.ts --runInBand --forceExit"
	},
	"author": "dao.xyz",
	"license": "MIT",
	"devDependencies": {
		"@types/libsodium-wrappers": "^0.7.10",
		"uint8arrays": "^4.0.4"
	},
	"dependencies": {
		"@peerbit/cache": "^2.0.4",
		"@dao-xyz/borsh": "^5.1.8",
		"@stablelib/sha256": "^1.0.1",
		"@peerbit/uint8arrays": "3.0.1",
		"@ethersproject/wallet": "^5.7.0",
		"@libp2p/crypto": "^3.0.3",
		"@libp2p/peer-id": "^4.0.3",
		"libsodium-wrappers": "^0.7.13"
	}
}
