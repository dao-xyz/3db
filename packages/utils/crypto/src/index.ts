export * from "./key.js";
export * from "./ed25519.js";
export * from "./signature.js";
export * from "./key.js";
export * from "./sepc256k1.js";
export * from "./x25519.js";
export * from "./encryption.js";
export * from "./libp2p.js";
export * from "./utils.js";
export * from "./hash.js";
export * from "./random.js";
export * from "./prehash.js";
export * from "./signer.js";
export * from "./keychain.js";
import libsodium from "libsodium-wrappers";
const ready = libsodium.ready; // TODO  can we export ready directly ?

export { ready };
