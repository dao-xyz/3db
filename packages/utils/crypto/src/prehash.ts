import { toUtf8Bytes } from "@ethersproject/strings";
import sha3 from "js-sha3";
import { concat } from "uint8arrays";
import { sha256 } from "./hash.js";

const messagePrefix = "\x19Ethereum Signed Message:\n";
const ethKeccak256Hash = (message: Uint8Array) =>
	new Uint8Array(
		sha3.keccak256
			.update(
				concat([
					toUtf8Bytes(messagePrefix),
					toUtf8Bytes(String(message.length)),
					message,
				]),
			)
			.arrayBuffer(),
	);

export enum PreHash {
	NONE = 0,
	SHA_256 = 1,
	// BLAKE3 = 2,
	ETH_KECCAK_256 = 3,
}

export const prehashFn = (
	data: Uint8Array,
	prehash: PreHash,
): Promise<Uint8Array> | Uint8Array => {
	if (prehash === PreHash.NONE) {
		return data;
	}
	if (prehash === PreHash.SHA_256) {
		return sha256(data);
	}
	if (prehash === PreHash.ETH_KECCAK_256) {
		return ethKeccak256Hash(data);
	}

	throw new Error("Unsupported");
};
