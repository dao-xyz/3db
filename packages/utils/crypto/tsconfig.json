{
	"extends": "aegir/src/config/tsconfig.aegir.json",
	"compilerOptions": {
		// allow top level await
		"experimentalDecorators": true,
		"emitDecoratorMetadata": true,
		"outDir": "dist"
	},
	"include": ["src", "test", "benchmark"]
}
