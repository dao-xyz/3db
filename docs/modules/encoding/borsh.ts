import {
	deserialize,
	field,
	fixedArray,
	serialize,
	variant,
} from "@dao-xyz/borsh";
import { randomBytes } from "@peerbit/crypto";

abstract class AbstractPost {}

@variant(0) // V0
class PostV0 extends AbstractPost {
	@field({ type: fixedArray("u8", 32) }) // id will always be a Uint8array with size 32
	id: Uint8Array;

	@field({ type: "string" })
	message: string;

	constructor(message: string) {
		super();
		this.id = randomBytes(32);
		this.message = message;
	}
}

@variant(1) // V1
class PostV1 extends AbstractPost {
	@field({ type: fixedArray("u8", 32) }) // id will always be a Uint8array with size 32
	id: Uint8Array;

	@field({ type: "string" })
	title: string;

	@field({ type: "string" })
	message: string;

	constructor(title: string, message: string) {
		super();
		this.id = randomBytes(32);
		this.title = title;
		this.message = message;
	}
}

const message = new PostV0("Hello world!");
const bytes = serialize(message); // [0, ... ] will start with 0 because @variant(0)
const post: AbstractPost = deserialize(bytes, AbstractPost);

if (post instanceof PostV0) {
	// To OLD behaviour
} else post instanceof PostV1;
{
	// Do V1 behaviour
}
