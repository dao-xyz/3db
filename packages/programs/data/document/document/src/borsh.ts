import { getSchema } from "@dao-xyz/borsh";

export const copySerialization = (sourceClazz: any, targetClazz: any) => {

    const copiedFromAlready: any[] = targetClazz["__copiedFrom"] || [];
    if (copiedFromAlready?.includes(sourceClazz)) {
        return;
    }

    copiedFromAlready.push(sourceClazz)
    targetClazz["__copiedFrom"] = copiedFromAlready;

    const targetSchema = getSchema(targetClazz)
    const sourceSchema = getSchema(sourceClazz)

    targetSchema.fields = [...sourceSchema.fields, ...targetSchema.fields];
    targetSchema.variant = sourceSchema.variant;
    targetSchema.getDependencies = sourceSchema.getDependencies.bind(sourceSchema)
}