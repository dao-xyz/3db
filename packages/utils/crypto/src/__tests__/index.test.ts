import {
	Ed25519Keypair,
	X25519Keypair,
	verifySignatureSecp256k1,
	Secp256k1PublicKey,
	Secp256k1Keypair,
	verify,
	Ed25519PublicKey,
	SignatureWithKey,
	Ed25519PrivateKey,
	toBase64
} from "../index.js";
import sodium from "libsodium-wrappers";
import { deserialize, serialize } from "@dao-xyz/borsh";
import { Wallet } from "@ethersproject/wallet";
import { createSecp256k1PeerId } from "@libp2p/peer-id-factory";
import { supportedKeys } from "@libp2p/crypto/keys";
import {
	verifySignatureEd25519,
	sign as signEd25519
} from "../ed25519-sign.js";
import {
	verifySignatureEd25519 as verifySignatureEd25519Browser,
	sign as signEd25519Browser
} from "../ed25519-sign-browser";
import { PreHash } from "../prehash.js";

describe("Ed25519", () => {
	it("ser/der", async () => {
		const keypair = await Ed25519Keypair.create();
		const bytes = serialize(keypair)
		const derser = deserialize(bytes, Ed25519Keypair);
		expect(new Uint8Array(derser.publicKey.publicKey)).toEqual(
			keypair.publicKey.publicKey
		);
	});

	it("PeerId", async () => {
		const kp = await Ed25519Keypair.create();
		const peerId = await kp.toPeerId();
		const privateKeyFromPeerId = Ed25519PrivateKey.fromPeerID(peerId);
		const keyPairFromPeerId = Ed25519Keypair.fromPeerId(peerId);
		expect(keyPairFromPeerId.equals(kp)).toBeTrue();
		expect(
			privateKeyFromPeerId.equals(keyPairFromPeerId.privateKey)
		).toBeTrue();
	});

	describe("native", () => {
		it("verify native", async () => {
			const keypair = sodium.crypto_sign_keypair();
			const data = new Uint8Array([1, 2, 3]);
			const signature = sodium.crypto_sign_detached(data, keypair.privateKey);
			const isVerified = await verifySignatureEd25519(
				new SignatureWithKey({
					prehash: PreHash.NONE,
					publicKey: new Ed25519PublicKey({ publicKey: keypair.publicKey }),
					signature: signature
				}),
				data
			);
			expect(isVerified).toBeTrue();
		});

		it("verify", async () => {
			const keypair = await Ed25519Keypair.create();
			const data = new Uint8Array([1, 2, 3]);
			const signature = await keypair.sign(data);
			const isVerified = await verifySignatureEd25519(signature, data);
			expect(isVerified).toBeTrue();

			const isNotVerified = await verify(signature, data.reverse());
			expect(isNotVerified).toBeFalse();
		});

		it("verify hashed", async () => {
			const keypair = await Ed25519Keypair.create();
			const data = new Uint8Array([1, 2, 3]);
			const signature = await keypair.sign(data, PreHash.SHA_256);
			const isVerified = await verify(signature, data);
			expect(isVerified).toBeTrue();

			const isNotVerified = await verifySignatureEd25519(
				signature,
				data.reverse()
			);
			expect(isNotVerified).toBeFalse();
		});
	});

	describe("browser", () => {
		it("verify", async () => {
			const keypair = await Ed25519Keypair.create();
			const data = new Uint8Array([1, 2, 3]);
			const signature = await signEd25519Browser(data, keypair, PreHash.NONE);
			const isVerified = await verifySignatureEd25519Browser(signature, data);
			expect(isVerified).toBeTrue();

			const isNotVerified = await verifySignatureEd25519Browser(
				signature,
				data.reverse()
			);
			expect(isNotVerified).toBeFalse();
		});

		it("verify hashed", async () => {
			const keypair = await Ed25519Keypair.create();
			const data = new Uint8Array([1, 2, 3]);
			const signature = await signEd25519Browser(
				data,
				keypair,
				PreHash.SHA_256
			);
			const isVerified = await verifySignatureEd25519Browser(signature, data);
			expect(isVerified).toBeTrue();

			const isNotVerified = await verifySignatureEd25519Browser(
				signature,
				data.reverse()
			);
			expect(isNotVerified).toBeFalse();
		});
	});

	describe("mixed api", () => {
		let signFns = [signEd25519Browser, signEd25519];
		let verifyFns = [verifySignatureEd25519Browser, verifySignatureEd25519];

		it("sign", async () => {
			const keypair = await Ed25519Keypair.create();
			const data = new Uint8Array([1, 2, 3]);
			let signatures: SignatureWithKey[] = [];
			for (const signFn of signFns) {
				const signature = await signFn(data, keypair, PreHash.NONE);
				for (const verifyFn of verifyFns) {
					expect(await verifyFn(signature, data)).toBeTrue();
				}

				signatures.push(signature);
				if (signatures.length > 1) {
					expect(
						signatures[signatures.length - 2].equals(
							signatures[signatures.length - 1]
						)
					).toBeTrue();
				}
			}
		});

		it("sign hashed", async () => {
			const keypair = await Ed25519Keypair.create();
			const data = new Uint8Array([1, 2, 3]);
			let signatures: SignatureWithKey[] = [];
			for (const fn of signFns) {
				const signature = await fn(data, keypair, PreHash.SHA_256);
				signatures.push(signature);

				for (const verifyFn of verifyFns) {
					expect(await verifyFn(signature, data)).toBeTrue();
				}

				if (signatures.length > 1) {
					expect(
						signatures[signatures.length - 2].equals(
							signatures[signatures.length - 1]
						)
					).toBeTrue();
				}
			}
		});
	});
});

describe("X25519", () => {
	it("ser/der", async () => {
		const keypair = await X25519Keypair.create();
		const derser = deserialize(serialize(keypair), X25519Keypair);
		expect(new Uint8Array(derser.publicKey.publicKey)).toEqual(
			keypair.publicKey.publicKey
		);
	});
});

describe("Sepck2561k1", () => {
	const data = new Uint8Array([1, 2, 3]);

	it("wallet sign", async () => {
		const wallet = Wallet.createRandom();
		const pk = await Secp256k1PublicKey.recover(wallet);
		const signature = await wallet.signMessage(data);
		let signatureBytes = Buffer.from(signature);

		const signatureWithKey = new SignatureWithKey({
			prehash: PreHash.ETH_KECCAK_256,
			publicKey: pk,
			signature: signatureBytes
		});

		const isVerified = await verifySignatureSecp256k1(signatureWithKey, data);
		expect(isVerified).toBeTrue();

		const isNotVerified = await verifySignatureSecp256k1(
			signatureWithKey,
			new Uint8Array(data).reverse()
		);
		expect(isNotVerified).toBeFalse();
	});
	it("keypair sign", async () => {
		const peerId = await createSecp256k1PeerId();
		const keypair = Secp256k1Keypair.fromPeerId(peerId);
		const privateKey = new supportedKeys["secp256k1"].Secp256k1PrivateKey(
			peerId.privateKey!.slice(4)
		);
		const publicKeyComputed = privateKey.public;
		expect(publicKeyComputed.bytes).toEqual(peerId.publicKey);
		const signature = await keypair.sign(data, PreHash.ETH_KECCAK_256);
		expect(await verifySignatureSecp256k1(signature, data)).toBeTrue();
	});

	it("PeerId", async () => {
		const kp = await Secp256k1Keypair.create();
		const peerId = await kp.toPeerId();
		expect(Secp256k1Keypair.fromPeerId(peerId).equals(kp));
	});

	it("ser/der", async () => {
		const wallet = await Wallet.createRandom();
		const pk = await Secp256k1PublicKey.recover(wallet);
		const derser = deserialize(serialize(pk), Secp256k1PublicKey);
		expect(derser.equals(pk)).toBeTrue();
	});
});
