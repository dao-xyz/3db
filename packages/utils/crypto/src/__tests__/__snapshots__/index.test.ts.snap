// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ed25519 ser/der 1`] = `"AADqtK4m5Nd0bR5zDCc/LnstxZuhg07H/CmvOYznbo+jswDqtK4m5Nd0bR5zDCc/LnstxZuhg07H/CmvOYznbo+jswB5tGeUT5sqWME2yA5ljox+scsrGjCv2Sm+PcZ0InQ5eg=="`;
